File: Random Marathon 68/G 17204.py

N, K = map(int, input().split())

list = [int(input()) for _ in range(N)]

visited = [False] * N

stack = [0]

cnt = 0

while stack:
    current = stack.pop()

    if current == K:
        print(cnt)
        break
    if visited[current]:
        print(-1)
        break

    cnt += 1
    stack.append(list[current])
    visited[current] = True

# 죨라 쉬운 dfs
        

File: Random Marathon 68/G 17204.c
#include <stdio.h>

int main(void)
{
    int N, K;
    scanf("%d %d", &N, &K);

    int arr[N];
    for (int i = 0; i < N; i++)
        scanf("%d", &arr[i]);

    int current = 0, next = 0;
    int cnt = 0;

    // K를 지목하거나 이미 지목한 사람을 지목하면 종료
    while (arr[current] != K && arr[current] != 0)
    {
        next = arr[current];
        arr[current] = 0;
        current = next;
        cnt++;
    }

    if (arr[current] == K)
        printf("%d", cnt + 1);
    else
        printf("-1");

    return 0;
}